### PR TITLE
add cluster_config_lock_ to avoid data race

### LINKED ISSUE #104: Data race error while modify config
### Description
We Implement a method `updateConfiguration` to update config (however, It‘s seem like only called during initialization). I guess it's for Dynamic Reconfiguration, but There is no complete implementation. 
```cpp
// RaftKeeper/src/Service/KeeperDispatcher.cpp
void KeeperDispatcher::updateConfiguration(const Poco::Util::AbstractConfiguration & config)
{
    auto diff = server->getConfigurationDiff(config);
    if (diff.empty())
        LOG_TRACE(log, "Configuration update triggered, but nothing changed for RAFT");
    else if (diff.size() > 1)
        LOG_WARNING(log, "Configuration changed for more than one server ({}) from cluster, it's strictly not recommended", diff.size());
    else
        LOG_INFO(log, "Configuration change size ({})", diff.size());

    for (auto & change : diff)
    {
        bool push_result = update_configuration_queue.push(change);
        if (!push_result)
            throw Exception(ErrorCodes::SYSTEM_ERROR, "Cannot push configuration update to queue");
    }
}

```
There is a background thread in nuraft::state_mgr that calls commit_conf to call NuRaftStateManager::save_config, which will modify cur_cluster_config, The main thread is calling updateConfiguration to read the modified cur_cluster_config, which will cause a data race

```cpp
// RaftKeeper/contrib/NuRaft/src/handle_commit.cxx
if (le->get_val_type() == log_val_type::conf) {
            commit_conf(index_to_commit, le);
        }
```

### How to reproduce 

found in https://github.com/JDRaftKeeper/RaftKeeper/actions/runs/6852423348/job/18630926726?pr=102

I can easily add a mutex for NuRaftStateManager::cur_cluster_config to fix it. But i got a lot of confusion for intention for `updateConfiguration`.

@nicelulu @JackyWoo cc

### PR DESCRIPTION
**Pull request by lzydmxy**
### Which issues of this PR fixes:

This PR try to fix #104

### Change log:
add cluster_config_lock_ to avoid data race

### CODE REVIEW

**Comment by lzydmxy** [comment]
@JackyWoo 2023.11.16 07:56:43.746655 [ 125 ] <Trace> ForwardingConnectionHandler: Forwarding handler socket readable
2023.11.16 07:56:43.746676 [ 125 ] <Trace> ForwardingConnectionHandler: forwarding handler socket available
2023.11.16 07:56:43.746689 [ 125 ] <Trace> ForwardingConnectionHandler: Try handle new package
2023.11.16 07:56:43.746708 [ 125 ] <Trace> ForwardingConnectionHandler: Receive Handshake
2023.11.16 07:56:43.746725 [ 125 ] <Trace> ForwardingConnectionHandler: forwarding handler socket available
2023.11.16 07:56:43.746898 [ 125 ] <Fatal> : Logical error: 'Server 2 client 9 already registered in dispatcher'.
2023.11.16 07:56:43.747143 [ 11 ] <Trace> BaseDaemon: Received signal 6
2023.11.16 07:56:43.747447 [ 130 ] <Fatal> BaseDaemon: ########################################
2023.11.16 07:56:43.747523 [ 130 ] <Fatal> BaseDaemon: (version v2.0.3, build id: 542805F1020432FC) (from thread 125) Received signal Aborted (6)
2023.11.16 07:56:43.747585 [ 130 ] <Fatal> BaseDaemon: 
2023.11.16 07:56:43.747677 [ 130 ] <Fatal> BaseDaemon: Stack trace: 0x7f793d529a7c 0x7f793d4d5476 0x7f793d4bb7f3 0x7fecde 0x7ff1d4 0xad9f59 0xac9e16 0xa5a555 0xa52f79 0xa6145d 0x1655184 0x9aaad0 0x9995f9 0x991dd9 0x1750429 0x174c652 0x7f793d527b43 0x7f793d5b8bb4

File: src/Service/NuRaftStateManager.h [hunk 1]
```diff
@@ -88,6 +90,12 @@ class NuRaftStateManager : public nuraft::state_mgr
 
     ptr<cluster_config> cur_cluster_config;
 
+    /**
+     * Lock for cluster config.
+     */
+    mutable std::mutex cluster_config_lock_;
```
Comment (JackyWoo) [inline src/Service/NuRaftStateManager.h:96]: cluster_config_lock_ -> cluster_config_mutex

File: src/Service/NuRaftStateManager.cpp [hunk 2]
```diff
@@ -24,22 +24,36 @@ NuRaftStateManager::NuRaftStateManager(int id_, const Poco::Util::AbstractConfig
...
+        return get_cluster_config();
     }
 
     std::unique_ptr<ReadBufferFromFile> read_file_buf = std::make_unique<ReadBufferFromFile>(cluster_config_file, 4096);
     size_t size;
     readVarUInt(size, *read_file_buf);
     ptr<nuraft::buffer> buf = nuraft::buffer::alloc(size);
     read_file_buf->readStrict(reinterpret_cast<char *>(buf->data()), size);
-    cur_cluster_config = nuraft::cluster_config::deserialize(*buf);
-    LOG_INFO(log, "load config with log index {}", cur_cluster_config->get_log_idx());
```
Comment (JackyWoo) [inline src/Service/NuRaftStateManager.cpp:41]: Why delete the logging?

File: src/Service/NuRaftStateManager.cpp [hunk 3]
```diff
@@ -24,22 +24,36 @@ NuRaftStateManager::NuRaftStateManager(int id_, const Poco::Util::AbstractConfig
...
 }
 
+ptr<cluster_config> NuRaftStateManager::get_cluster_config() const
+{
+    std::lock_guard<std::mutex> l(cluster_config_lock_);
+    ptr<cluster_config> ret = cur_cluster_config;
+    return ret;
+}
+
+void NuRaftStateManager::set_cluster_config(const ptr<cluster_config>& new_config)
```
Comment (JackyWoo) [inline src/Service/NuRaftStateManager.cpp:34]: set_cluster_config -> setClusterConfig

File: src/Service/NuRaftStateManager.cpp [hunk 4]
```diff
@@ -24,22 +24,36 @@ NuRaftStateManager::NuRaftStateManager(int id_, const Poco::Util::AbstractConfig
...
+{
+    std::lock_guard<std::mutex> l(cluster_config_lock_);
```
Comment (JackyWoo) [inline src/Service/NuRaftStateManager.cpp:36]: veriable `l` is too short

**Comment by JackyWoo** [comment]
The error may not caulsed by your changing, I will take a look at it latter.

File: src/Service/NuRaftStateManager.cpp (hunk 2 above)
Comment (lzydmxy) [reply src/Service/NuRaftStateManager.cpp:41]: I just move log into function `setClusterConfig`, maybe it's better to print log after call it.

**Comment by JackyWoo** [comment]
@lzydmxy please rebase the recent master branch

**Comment by JackyWoo** [comment]
Mark the [error test ](https://github.com/JDRaftKeeper/RaftKeeper/actions/runs/6901574717/job/18776625901?pr=106)

**Comment by lzydmxy** [comment]
> @lzydmxy please rebase the recent master branch

okay

**Comment by lzydmxy** [comment]
> Mark the [error test ](https://github.com/JDRaftKeeper/RaftKeeper/actions/runs/6901574717/job/18776625901?pr=106)

closed in #110

**Pull request merged**